zephyr/Kconfig

```diff
@@ -0,0 +1,680 @@
+	default 1
+	range 1 1 if SINGLE_APPLICATION_SLOT
+	help
+	  Enables support of multi image update.
+
+choice BOOT_DOWNGRADE_PREVENTION_CHOICE
+	prompt "Downgrade prevention"
+	optional
+
+config MCUBOOT_DOWNGRADE_PREVENTION
+	bool "SW based downgrade prevention"
+	depends on !BOOT_DIRECT_XIP
+	help
+	  Prevent downgrades by enforcing incrementing version numbers.
+	  When this option is set, any upgrade must have greater major version
+	  or greater minor version with equal major version. This mechanism
+	  only protects against some attacks against version downgrades (for
+	  example, a JTAG could be used to write an older version).
+
+config MCUBOOT_DOWNGRADE_PREVENTION_SECURITY_COUNTER
+	bool "Use image security counter instead of version number"
+	depends on MCUBOOT_DOWNGRADE_PREVENTION
+	depends on (BOOT_SWAP_USING_MOVE || BOOT_SWAP_USING_SCRATCH)
+	help
+       Security counter is used for version eligibility check instead of pure
+       version.  When this option is set, any upgrade must have greater or
+       equal security counter value.
+       Because of the acceptance of equal values it allows for software
+       downgrades to some extent.
+
+config MCUBOOT_HW_DOWNGRADE_PREVENTION
+	bool "HW based downgrade prevention"
+	help
+	  Prevent undesirable/malicious software downgrades. When this option is
+	  set, any upgrade must have greater or equal security counter value.
+	  Because of the acceptance of equal values it allows for software
+	  downgrade to some extent.
+
+endchoice
+
+config BOOT_WATCHDOG_FEED
+	bool "Feed the watchdog while doing swap"
+	default y if WATCHDOG
+	default y if SOC_FAMILY_NRF
+	# for nRF nrfx based implementation is available
+	imply NRFX_WDT if SOC_FAMILY_NRF
+	imply NRFX_WDT0 if SOC_FAMILY_NRF
+	imply NRFX_WDT1 if SOC_FAMILY_NRF
+	help
+	  Enables implementation of MCUBOOT_WATCHDOG_FEED() macro which is
+	  used to feed watchdog while doing time consuming operations.
+
+config BOOT_IMAGE_ACCESS_HOOKS
+	bool "Enable hooks for overriding MCUboot's native routines"
+	help
+	  Allow to provide procedures for override or extend native
+	  MCUboot's routines required for access the image data and the image
+	  update. It is up to the project customization to add required source
+	  files to the build.
+
+config MCUBOOT_ACTION_HOOKS
+	bool "Enable hooks for responding to MCUboot status changes"
+	help
+	  This will call a handler when the MCUboot status changes which allows
+	  for some level of user feedback, for instance to change LED status to
+	  indicate a failure, using the callback:
+	  'void mcuboot_status_change(mcuboot_status_type_t status)' where
+	  'mcuboot_status_type_t' is listed in
+	  boot/bootutil/include/bootutil/mcuboot_status.h
+
+endmenu
+
+config MCUBOOT_DEVICE_SETTINGS
+	# Hidden selector for device-specific settings
+	bool "Non-hidden selector for device-specific settings"
+	default n
+        # CPU options
+	select MCUBOOT_DEVICE_CPU_CORTEX_M0 if CPU_CORTEX_M0
```

Comment:
this should be kept as automatic hidden selection  as it used to be.

Reply:
@nvlsianpu 
When relocating the Kconfig definitions from `$mcuboot/boot/zephyr/Kconfig` (project-specific) to `$mcuboot/zephyr/Kconfig` (module-specific), the definition of `MCUBOOT_DEVICE_SETTINGS` is now visible to the `$zephyr/samples/tfm_integration/psa-crypto/` build. This build with existing default 'y' now fails with:

> warning: FLASH_MAP (defined at subsys/storage/flash_map/Kconfig:10) has direct dependencies FLASH_HAS_DRIVER_ENABLED with value n, but is currently being y-selected by the following symbols:
> - MCUBOOT_DEVICE_SETTINGS (defined at /home/zephyr/mcubbootwksp/mcuboot/zephyr/Kconfig:620), with value y, direct dependencies y (value: y)

In other words, refactoring these Kconfigs for reuse also exposes them to more existing projects. Defaulting anything to 'y' impacts those projects too.  

AFAICT to not effect existing projects the default needs to be 'n' and the control point (e.g., `MCUBOOT_DEVICE_SETTINGS`) needs to be selectable by the projects that need it.  Do you see a better solution?